weave/trace_server/execution_runner/trace_server_adapter.py

```diff
@@ -0,0 +1,250 @@
+"""
+Trace server adapter module for secure user context injection.
+
+This module provides adapters that wrap trace servers to enforce user and project
+scoping. It's a critical security component that ensures users can only access
+their own data by converting between internal and external references.
+
+Key components:
+- externalize_trace_server: Main function to wrap a trace server with security
+- IdConverter: Validates and converts project/user IDs
+- UserInjectingExternalTraceServer: Ensures user ID is injected into all requests
+
+Security model:
+- All project IDs are prefixed with SERVER_SIDE_ENTITY_PLACEHOLDER when externalized
+- Project ID validation prevents cross-project access
+- User ID injection ensures all operations are scoped to the authenticated user
+- Run IDs are explicitly not supported for server-side evaluation
+"""
+
+from __future__ import annotations
+
+from typing import Callable, TypeVar
+
+from pydantic import BaseModel
+
+from weave.trace_server import external_to_internal_trace_server_adapter
+from weave.trace_server import trace_server_interface as tsi
+from weave.trace_server.trace_server_converter import universal_int_to_ext_ref_converter
+
+# Special placeholder used to indicate server-managed entities
+# This prevents clients from specifying arbitrary entities
+SERVER_SIDE_ENTITY_PLACEHOLDER = "__SERVER__"
+SERVER_SIDE_PROJECT_ID_PREFIX = SERVER_SIDE_ENTITY_PLACEHOLDER + "/"
+
+
+def externalize_trace_server(
```

Comment:
nit: prefer these "optional wrap" functions to start with `with`, which makes it easier to skim what's going on, e.g.

```py
server = with_feature1(with_feature2(...))
```

Reply:
changed